core/lib/types/src/fee_model.rs

```diff
@@ -0,0 +1,179 @@
+pub enum BatchFeeInput {
+    L1Pegged(L1PeggedBatchFeeModelInput),
+    PubdataIndependent(PubdataIndependentBatchFeeModelInput),
+}
+
+impl BatchFeeInput {
+    pub fn zero() -> Self {
+        Self::L1Pegged(L1PeggedBatchFeeModelInput {
+            l1_gas_price: 0,
+            fair_l2_gas_price: 0,
+        })
+    }
+}
+
+impl Default for BatchFeeInput {
+    fn default() -> Self {
+        // We have a sensible default value of 1 gwei for L1 gas price and 0.1 gwei for fair L2 gas price.
+        Self::L1Pegged(L1PeggedBatchFeeModelInput {
+            l1_gas_price: 1_000_000_000,
+            fair_l2_gas_price: 100_000_000,
+        })
+    }
+}
+
+impl BatchFeeInput {
+    pub fn into_l1_pegged(self) -> L1PeggedBatchFeeModelInput {
+        match self {
+            BatchFeeInput::L1Pegged(input) => input,
+            _ => panic!("Can not convert PubdataIndependentBatchFeeModelInput into L1PeggedBatchFeeModelInput"),
+        }
+    }
+
+    pub fn l1_pegged_ref_mut(&mut self) -> &mut L1PeggedBatchFeeModelInput {
+        match self {
+            BatchFeeInput::L1Pegged(input) => input,
+            _ => panic!("Can not convert PubdataIndependentBatchFeeModelInput into L1PeggedBatchFeeModelInput"),
+        }
+    }
+
+    pub fn l1_pegged_ref(&self) -> &L1PeggedBatchFeeModelInput {
+        match self {
+            BatchFeeInput::L1Pegged(input) => input,
+            _ => panic!("Can not convert PubdataIndependentBatchFeeModelInput into L1PeggedBatchFeeModelInput"),
+        }
+    }
+
+    pub fn fair_pubdata_price(&self) -> u64 {
+        match self {
+            BatchFeeInput::L1Pegged(input) => input.l1_gas_price * L1_GAS_PER_PUBDATA_BYTE as u64,
+            BatchFeeInput::PubdataIndependent(input) => input.fair_pubdata_price,
+        }
+    }
+
+    pub fn fair_l2_gas_price(&self) -> u64 {
+        match self {
+            BatchFeeInput::L1Pegged(input) => input.fair_l2_gas_price,
+            BatchFeeInput::PubdataIndependent(input) => input.fair_l2_gas_price,
+        }
+    }
+
+    pub fn l1_gas_price(&self) -> u64 {
+        match self {
+            BatchFeeInput::L1Pegged(input) => input.l1_gas_price,
+            BatchFeeInput::PubdataIndependent(input) => input.l1_gas_price,
+        }
+    }
+
+    pub fn into_pubdata_independent(self) -> PubdataIndependentBatchFeeModelInput {
+        match self {
+            BatchFeeInput::PubdataIndependent(input) => input,
+            BatchFeeInput::L1Pegged(input) => PubdataIndependentBatchFeeModelInput {
+                fair_l2_gas_price: input.fair_l2_gas_price,
+                fair_pubdata_price: input.l1_gas_price * 17,
+                l1_gas_price: input.l1_gas_price,
+            },
+        }
+    }
+
+    pub fn for_protocol_version(
+        protocol_version: ProtocolVersionId,
+        fair_l2_gas_price: u64,
+        fair_pubdata_price: Option<u64>,
+        l1_gas_price: u64,
+    ) -> Self {
+        if protocol_version.is_1_4_1() {
```

Comment:
this is a little bit dangerous - if we roll out 1.5.0 - then this will return 'false' - and we'll switch back to L1Pegged.

Reply:
I changed to `is_post_1_4_1`